packages/flutter_tools/lib/src/android/adb.dart

```diff
@@ -112,6 +118,26 @@ class Adb {
     return controller.stream;
   }
 
+  Future _populateDeviceNames(List<AdbDevice> devices) async {
+    for (AdbDevice device in devices) {
+      if (device.modelID == null) {
+        // If we don't have a name of a device in our cache, call `device -l` to populate it.
+        if (_idToNameCache[device.id] == null)
+          await _populateDeviceCache();
+
+        // Set the device name from the cached name. Adb device notifications only
+        // have IDs, not names. We get the name by calling `listDevices()`.
+        device.modelID = _idToNameCache[device.id];
+      }
+    }
+  }
+
+  Future _populateDeviceCache() async {
+    List<AdbDevice> devices = await listDevices();
+    for (AdbDevice device in devices)
+      _idToNameCache[device.id] = device.modelID;
+  }
+
```

Comment:
This is racy. Suppose you had devices A and B but not C, then you populate the names of A and B, then for C you repopulate the entire list, which might update A and B, after we've set their modelIDs.

It seems better to have _idToNameCache be null until _populateDeviceCache populates the whole thing, and then once it does that, assume it's been done, or something. Even better would be if adb could give us a holistic view of all the information we need in one command, I guess.